### PR TITLE
ci(*:skip) Add Python script to check PR titles

### PR DESCRIPTION
**Pull request by charlesbvll**
### Description

We want to enforce the following PR title format:

```
<type>(<project>) <subject>
```

(or `<type>(<project>:skip) <subject>` to ignore the PR in the changelog)

Where `<type>` $\in$ `{ci, fix, feat, docs, refactor, break}`, `<project>` $\in$ `{framework, baselines, datasets, examples, '*'}`, and `<subject>` starts with a capitalised verb in the imperative mood.

Valid examples:

- `feat(framework) Add flwr build CLI command`
- `refactor(examples:skip) Improve quickstart-pytorch logging`
- `ci(*:skip) Enforce PR title format`

Invalid examples:

- `feat(framework): Add flwr build CLI command` (extra `:`)
- `feat(skip) Add flwr build CLI command` (missing `<project>`)
- `feat(framework) add flwr build CLI command` (non capitalised verb)
- `feat(framework) Add flwr build CLI command.` (dot at the end)
- `Add flwr build CLI command` (missing `<type>(<project>)`)

### Related issues/PRs

N/A

### Explanation

- [x] Create a configuration file defining the defined rules
- [x] Write a Python script to check if a title has the correct format using this config file
- [x] Add a CI check to enforce the new format using this script, the CI should fail if the format is not respected.
- [x] Update the contributor docs

### Checklist

- [x] Implement proposed change
- [x] Write tests
- [x] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [x] Update the changelog entry below
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Changelog entry

<skip>

### Any other comments?

N/A

### CODE REVIEW

File: dev/changelog.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+types:
```
Comment (jafermarq) [inline dev/changelog.yml:1]: Should we leave a comment at the top of the file to just indicate it's by `check_pr_title.py`?

Comment (danieljanes) [inline dev/changelog.yml:1]: For consistency since `project` is also singular:
```suggestion
type:
```

File: dev/check_pr_title.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    pr_title = sys.argv[1]
+
+    # Load the YAML configuration
+    with (pathlib.Path(__file__).parent.resolve() / "changelog.yml").open(
+        encoding="utf-8"
+    ) as file:
+        config = yaml.safe_load(file)
+
+    # Extract types, project, and scope from the config
+    types = "|".join(config["types"])
```
Comment (danieljanes) [inline dev/check_pr_title.py:35]:
```suggestion
    types = "|".join(config["type"])
```

File: dev/check_pr_title.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    scope = config["scope"]
+
+    # Construct the pattern
+    pattern_template = config["pattern_template"]
+    pattern = pattern_template.format(types=types, project=project, scope=scope)
+
+    # Check for the pattern in the first argument given to the script
+    if re.search(pattern, pr_title):
+        print("PR title is valid")
+        sys.exit(0)
```
Comment (jafermarq) [inline dev/check_pr_title.py:46]: Should we do before the `sys.exit(0)` the check for "imperative mood"

File: dev/check_pr_title.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    else:
+        print(
+            f"PR title `{pr_title}` is invalid, it should be of the form: <PR_TYPE>(<PR_SCOPE>) "
```
Comment (danieljanes) [inline dev/check_pr_title.py:49]:
```suggestion
            f"PR title `{pr_title}` is invalid, it should be of the form: <PR_TYPE>(<PR_PROJECT>) "
```

File: dev/check_pr_title.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            f"<PR_SUBJECT> with <PR_TYPE> in {types}, and "
+            f"<PR_SCOPE> in {project} where '*' is used when modifying multiple projects),"
```
Comment (danieljanes) [inline dev/check_pr_title.py:51]:
```suggestion
            f"<PR_PROJECT> in {project} where '*' is used when modifying multiple projects),"
```

File: dev/check_pr_title.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            "and <PR_SUBJECT> starting with "
+            "a capitalized verb in the imperative mood and without a dot at the end.\n"
```
Comment (danieljanes) [inline dev/check_pr_title.py:53]:
```suggestion
            "a capitalized verb in the imperative mood and without any punctuation at the end.\n"
```

File: doc/source/contributor-tutorial-contribute-on-github.rst [hunk 7]
```diff
@@ -304,39 +305,11 @@ Open PR
...
-
-Here are a few positive examples which provide helpful information without repeating how they do it, as that is already visible in the "Files changed" section of the PR:
-
-* Update docs banner to mention Flower Summit 2023
-* Remove unnecessary XGBoost dependency
-* Remove redundant attributes in strategies subclassing FedAvg
-* Add CI job to deploy the staging system when the ``main`` branch changes
-* Add new amazing library which will be used to improve the simulation engine
-
-
```
Comment (danieljanes) [inline doc/source/contributor-tutorial-contribute-on-github.rst:339]: Should we keep this and update it to be in line with the new format?

Comment (charlesbvll) [reply doc/source/contributor-tutorial-contribute-on-github.rst:339]: It felt a bit redondant with the appendix

File: .github/workflows/pr_check.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  cancel-in-progress: true
+
+jobs:
+  pr_title_check:
+    runs-on: ubuntu-22.04
+    name: Title format
+    steps:
+      - uses: actions/checkout@v4
+      - name: Bootstrap
+        uses: ./.github/actions/bootstrap
```
Comment (Robert-Steiner) [inline .github/workflows/pr_check.yml:20]: if we only need Python and pip, we could also use the official Python action: https://github.com/actions/setup-python.

Comment (charlesbvll) [reply .github/workflows/pr_check.yml:20]: I used the bootstrap action to make sure that if we need to update the Python version, it is all in the same place

File: dev/check_pr_title.py (hunk 3 above)
Comment (charlesbvll) [reply dev/check_pr_title.py:48]: @jafermarq I added a check for imperative mood, @danieljanes I placed the list of allowed verbs inside `changelog.yml`, is that good?

Comment (jafermarq) [reply dev/check_pr_title.py:48]: fine with me

File: doc/source/contributor-tutorial-contribute-on-github.rst (hunk 7 above)
Comment (jafermarq) [reply doc/source/contributor-tutorial-contribute-on-github.rst:339]: I also feel it's not necessary to keep

File: dev/check_pr_title.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
```
Comment (tanertopal) [inline dev/check_pr_title.py:1]:
```suggestion
#!/usr/bin/env python
# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
```

Should we make this dev script useable in the same way all other dev scripts are used?

Adding the shebang and making the script executable would turn

```
python ./dev/check_pr_title.py "${{ github.event.pull_request.title }}"
```
into
```
./dev/check_pr_title.py "${{ github.event.pull_request.title }}"
```

Comment (danieljanes) [reply dev/check_pr_title.py:1]: Ok for me to do it or not. Independent of this, let's merge the PR asap.

Comment (charlesbvll) [reply dev/check_pr_title.py:1]: I think we can ignore this for now, as it only makes sense if we have no dependencies associated with the script. I think we can merge the PR

Comment (danieljanes) [reply dev/check_pr_title.py:1]: @tanertopal btw, it won't be necessary for anyone to run this manually

Comment (tanertopal) [reply dev/check_pr_title.py:1]: Got it. Let's merge it then and improve later. I personally would like to remove the YAML dependency and replace it with TOML (therefore use Py3.11 for this job) but all that can be done afterwards so this is already applied as you two suggested.

**Review by tanertopal — APPROVED** [review]
lgtm!

**Pull request merged**